ci/codecov.sh

```diff
@@ -23,51 +23,69 @@ set -ex
 if [[ "$1" == "setup" ]]; then
     export B2_VARIANT=debug
     if [ -z "$B2_CI_VERSION" ]; then
-        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }cxxflags=-fkeep-static-functions cxxflags=--coverage"
+        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }cxxflags=-fprofile-arcs cxxflags=-ftest-coverage cxxflags=--coverage"
     else
-        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }-fkeep-static-functions --coverage"
+        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }-fprofile-arcs -ftest-coverage --coverage"
     fi
     export B2_LINKFLAGS="${B2_LINKFLAGS:+$B2_LINKFLAGS }--coverage"
 
 elif [[ "$1" == "upload" ]]; then
     if [ -z "$GCOV" ]; then
-        ver=7 # default
-        if [ "${B2_TOOLSET%%-*}" == "gcc" ]; then
-            if [[ "$B2_TOOLSET" =~ gcc- ]]; then
-                ver="${B2_TOOLSET##*gcc-}"
-            elif [[ "$B2_COMPILER" =~ gcc- ]]; then
-                ver="${B2_COMPILER##*gcc-}"
+        # enforce.sh ensures B2_COMPILER
+        COMPILER=${B2_COMPILER}
+        COMPILER_FAMILY=$(echo ${COMPILER} | cut -d'-' -f1)
+        COMPILER_EDITION=$(echo ${COMPILER} | cut -d'-' -f2)
+        # clang is still not working - need a llvm-cov wrapper and fixup pathing
+        if [ "${COMPILER_FAMILY}" == "clang" ]; then
+            echo "clang is not supported for coverage yet"
+            exit 1
+            # find /usr/lib -name 'llvm-cov*' -print
+            # LLVM_COV_PATH=$(find /usr/lib -type f -name llvm-cov -exec dirname {} \;)
+            # export PATH=${LLVM_COV_PATH}:${PATH}
+            # GCOV=llvm-cov
+        elif [ "${COMPILER_FAMILY}" == "gcc" ]; then
+            # cutting "gcc" above yields "gcc" for both values if no version present
+            if [ "${COMPILER_EDITION}" == "${COMPILER_FAMILY}" ]; then
+                GCOV=gcov
+            else
+                GCOV=gcov-${COMPILER_EDITION}
             fi
+        else
+            echo "Cannot determine GCOV."
+            exit 1
         fi
-        GCOV=gcov-${ver}
     fi
 
     # install the latest lcov we know works
     rm -rf /tmp/lcov
     cd /tmp
-    git clone --depth 1 -b v1.14 https://github.com/linux-test-project/lcov.git
+    git clone --depth 1 -b v1.15 https://github.com/linux-test-project/lcov.git
     export PATH=/tmp/lcov/bin:$PATH
     command -v lcov
     lcov --version
 
     # switch back to the original source code directory
     cd $BOOST_CI_SRC_FOLDER
-    : "${LCOV_BRANCH_COVERAGE:=1}" # Set default
+    : "${LCOV_BRANCH_COVERAGE:=1}" # On by default, job can override but advise against it
```

Comment:
```suggestion
    : "${LCOV_BRANCH_COVERAGE:=1}" # On by default, can be overridden in the CI config
```

I don't understand the "but advise against it" part. I can guess what it means, but it sounds strange. I also don't agree with that. Especially in C++ with exceptions etc. there are to many hidden branches which IMO make branch coverage to much of a hassle to deal with.
Anyway: I'd not add an opinion here.

Reply:
Fine, I will remove it, but I firmly believe that branch testing is the only accurate coverage number and if I had my way I would force everyone to use it.